File: server/common/protocol.py
DELIMITER = ","
BET_SEPARATOR = "\n"

OP_OK = 0
OP_BET = 1
OP_ERR = 2

HEADER_LEN = 4
OPCODE_LEN = 1


class Protocol:
    def __init__(self, sock):
        """Initialize the Protocol with a socket-like object."""
        self.sock = sock

    def _recv_all(self, length):
        """Read exactly `length` bytes from the socket and return them as bytes"""
        data = b""
        while len(data) < length:
            chunk = self.sock.recv(length - len(data))
            if not chunk:
                raise ConnectionError("socket closed")
            data += chunk
        return data

    def receive(self):
        """Receive a single framed packet and return its opcode and payload as a tuple."""
        opcode_bytes = self._recv_all(OPCODE_LEN)
        opcode = opcode_bytes[0]

        header = self._recv_all(HEADER_LEN)
        length = int.from_bytes(header, byteorder="big", signed=False)

        payload = ""

        if length > 0:
            payload = self._recv_all(length).decode("utf-8")

        return opcode, payload

    def send(self, opcode: int, message: str = ""):
        """Send a framed packet with the given opcode and UTF-8 payload."""
        data = message.encode("utf-8")

        header = len(data).to_bytes(HEADER_LEN, "big")

        packet = bytes([opcode]) + header + data
        self.sock.sendall(packet)

    def receive_message(self):
        """Receive a batch of bets and return them as a list of Bet objects"""
        opcode, payload = self.receive()

        if opcode != OP_BET:
            raise ValueError("unknown opcode")

        bets = []

        lines = payload.strip().split(BET_SEPARATOR)

        for line in lines:
            bet = line.split(DELIMITER)
            bets.append(bet)

        return bets

File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"net"
	"strings"
)

const (
	OpOK  byte = 0
	OpBet byte = 1
	OpErr byte = 2

	OpcodeLen = 1
	HeaderLen = 4

	Delimiter   = ","
	BetSeparator = "\n"
)

type Protocol struct {
	conn net.Conn
}

// NewProtocol returns a new Protocol that wraps the given network connection.
func NewProtocol(conn net.Conn) *Protocol {
	return &Protocol{conn: conn}
}

// Sends a framed packet to the Server.
// Returns an error if any write fails.
func (p *Protocol) send(opcode byte, message string) error {
	data := []byte(message)

	length := make([]byte, HeaderLen)
	binary.BigEndian.PutUint32(length, uint32(len(data)))

	packet := append([]byte{opcode}, length...)
	packet = append(packet, data...)

	total := 0
	for total < len(packet) {
		n, err := p.conn.Write(packet[total:])
		if err != nil {
			return err
		}
		total += n
	}

	return nil
}

// recvAll reads exactly length bytes from the Server, returning the data it read.
func (p *Protocol) recvAll(length int) ([]byte, error) {
	data := make([]byte, length)
	total := 0

	for total < length {
		n, err := p.conn.Read(data[total:])
		if err != nil {
			return nil, err
		}
		total += n
	}

	return data, nil
}

// receive reads a single framed packet from the Server.
// It returns the opcode, payload string and an error if one occurred.
func (p *Protocol) receive() (byte, string, error) {

	opcodeBytes, err := p.recvAll(OpcodeLen)
	if err != nil {
		return 0, "", err
	}

	opcode := opcodeBytes[0]

	header, err := p.recvAll(HeaderLen)
	if err != nil {
		return 0, "", err
	}

	length := binary.BigEndian.Uint32(header)

	payload, err := p.recvAll(int(length))
	if err != nil {
		return 0, "", err
	}

	return opcode, string(payload), nil
}

// SendBet builds a bet payload from the provided fields and sends it using the OpBet opcode.
func (p *Protocol) SendBet(
	agency string,
	nombre string,
	apellido string,
	dni string,
	nacimiento string,
	numero string,
) error {

	parts := []string{agency, nombre, apellido, dni, nacimiento, numero}
	payload := strings.Join(parts, Delimiter)

	return p.send(OpBet, payload)
}

func (p *Protocol) SendBatch(bets []string) error {

	payload := strings.Join(bets, BetSeparator)

	return p.send(OpBet, payload)
}